mantidproject/mantid
# LoadEventNexus code cleanup and small performance improvements

### Description of work

It was hypothesized that an appreciable amount of time for `LoadEventNexus` was spent in the NeXus-api calling the hdf5-api. as seen in this picture from vtune (bigger file, local laptop)

![image (1)](https://github.com/mantidproject/mantid/assets/404003/80afc7bd-3b03-4b48-9877-761914794022)

This appears to not be the case. 

With that information, this PR aims to reduce copying objects in `LoadBankFromDiskTask`, refactors part of `ProcessBankData` to make the pre-counting events more visible in profiling, and reworks part of `BankPulseTime` to read parts of the file at a time and convert those parts. This last part is a large portion of the speed improvements seen.

#### Timings

The files being used are (both in `/SNS/SNAP/IPTS-28913/` subtree)
* `SNAP_57514.lite.nxs.h5` 7.7GB file, 10.6GB workspace
* `SNAP_57480.nxs.h5` 21GB file, 29.8GB workspace

The baseline for "fastest possible" is using `dd` to copy the file to `/dev/null`. 
This makes the time for reading and not writing.
`LoadEventNexus` is run with only a filename and output workspace name.

Timing on my laptop (run 5 times each):

| file                  | dd    | `main`   | this PR |
|--------------------|------|-------|----------|
| SNAP_57514 |  6.1s | 6.8s | 6.5s |
| SNAP_57480 | 17.0s | 28.2s  | 26.5s |

Timing on my analysis-node17 (run 5 times each for little file, run 2 times each for big file):

| file                 | dd    | `main`   | this PR |
|-------------------|-------|-------|----------|
| SNAP_57514 |    23s | 1m21s | 1m36s |
| SNAP_57480 | 1m4s | 4m20s | 4m55s |

analysis computer times for `LoadEventNexus` varied by ~30s

*There is no associated issue,* but it is related to [EWM2463](https://ornlrse.clm.ibmcloud.com/ccm/web/projects/Neutron%20Data%20Project%20(Change%20Management)#action=com.ibm.team.workitem.viewWorkItem&id=2463)

### To test:

Since this is performance improvement, there isn't much to test. Try loading event nexus files and see that they still work.

*This does not require release notes* because it is a very minor performance improvement and is more of a code cleanup.





---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.
